File: linked-list/61RotateList.java
public class RotateList {
    class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public ListNode rotateRight(ListNode head, int k) {
        if(head == null || head.next == null) return head;
        int n = 1;
        ListNode oldHead = head;
        while(oldHead.next != null) {
            oldHead = oldHead.next;
            n++;
        }

        ListNode new_tail = head, new_head;
        for(int i = 0; i < n - k % n - 1; i++) {
            new_tail = new_tail.next;
        }
        oldHead.next = head;
        new_head = new_tail.next;
        new_tail.next = null;
        return new_head;
    }

    public ListNode rotateLeft(ListNode head, int k) {
        return null;
    }

}

File: amazon/medium/61RotateList.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        len_ll = 0
        curr_node = head
        old_tail = None
        while curr_node.next:
            len_ll += 1
            curr_node = curr_node.next
            old_tail = curr_node
        
        k = k % len_ll
        if k == 0:
            return head
        new_tail_count = len_ll - k - 2
        old_head = head
        curr_node = head
        while curr_node:
            if new_tail_count == 0:
                break
            curr_node = curr_node.next
            new_tail_count -= 1
        new_tail = curr_node
        new_head = curr_node.next
        new_tail.next = None
        old_tail.next = head
        return new_head
        
        